zipline/pipeline/loaders/blaze.py

```diff
@@ -0,0 +1,904 @@
+
+    Returns
+    -------
+    adjustments : dict[idx -> Float64Overwrite]
+        The adjustments dictionary to feed to the adjusted array.
+    """
+    ad_series = deltas[AD_FIELD_NAME]
+    adjustments = defaultdict(list)
+    for sid_idx, (sid, per_sid) in enumerate(deltas[column_name].iteritems()):
+        for kd, v in per_sid.iteritems():
+            adjustments[dates.get_loc(kd)].extend(
+                overwrite_from_dates(
+                    ad_series.loc[kd, sid],
+                    dates,
+                    dense_dates,
+                    sid_idx,
+                    v,
+                ),
+            )
+    return dict(adjustments)  # no subclasses of dict
+
+
+class BlazeLoader(dict):
+    def __init__(self, colmap=None):
+        self.update(colmap or {})
+
+    @classmethod
+    @memoize(cache=WeakKeyDictionary())
+    def global_instance(cls):
+        return cls()
+
+    def __hash__(self):
+        return id(self)
+
+    def __call__(self, column):
+        if column.dataset in self:
+            return self
+        raise KeyError(column)
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        return map(
+            op.getitem(
+                dict(concat(
+                    self._load_dataset(cs, dates, assets, mask)
+                    for _, cs in iteritems(groupby(getdataset, columns))
+                )),
+            ),
+            columns,
+        )
+
+    def _load_dataset(self, columns, dates, assets, mask):
+        try:
+            dataset, = set(map(getdataset, columns))
+        except ValueError:
+            raise AssertionError('all columns must come from the same dataset')
+
+        expr, deltas, resources = self[dataset]
+        have_sids = SID_FIELD_NAME in expr.fields
+        assets = list(map(int, assets))  # coerce from numpy.int64
+        fields = list(map(dataset_name, columns))
+        query_fields = fields + [AD_FIELD_NAME, TS_FIELD_NAME] + (
+            [SID_FIELD_NAME] if have_sids else []
+        )
+
+        def where(e):
+            """Create the query to run against the resources.
+
+            Parameters
+            ----------
+            e : Expr
+                The baseline or deltas expression.
+
+            Returns
+            -------
+            q : Expr
+                The query to run.
+            """
+            ts = e[TS_FIELD_NAME]
+            # Hack to get the lower bound to query:
+            # This must be strictly executed because the data for `ts` will
+            # be removed from scope too early otherwise.
+            lower = odo(ts[ts <= dates[0]].max(), pd.Timestamp)
+            selection = ts <= dates[-1]
+            if have_sids:
+                selection &= e[SID_FIELD_NAME].isin(assets)
+            if lower is not pd.NaT:
+                selection &= ts >= lower
+
+            return e[selection][query_fields]
+
+        extra_kwargs = {'d': resources} if resources else {}
```

Comment:
If you pass a bound expression, then resources is None, if you pass unbound expressions, then you need to pass data. `from_blaze` makes sure that data came in from either a bound expression or resources but not both.